src/App/Metadata.cpp

```diff
@@ -0,0 +1,750 @@
+namespace fs = boost::filesystem;
+XERCES_CPP_NAMESPACE_USE
+
+
+Metadata::Metadata(const fs::path& metadataFile)
+{
+    // Any exception thrown by the XML code propagates out and prevents object creation
+    XMLPlatformUtils::Initialize();
+
+    _parser = std::make_shared<XercesDOMParser>();
+    _parser->setValidationScheme(XercesDOMParser::Val_Never);
+    _parser->setDoNamespaces(true);
+
+    auto errHandler = std::make_unique<HandlerBase>();
+    _parser->setErrorHandler(errHandler.get());
+
+    _parser->parse(metadataFile.string().c_str());
+
+    auto doc = _parser->getDocument();
+    _dom = doc->getDocumentElement();
+
+    auto rootTagName = StrXUTF8(_dom->getTagName()).str;
+    if (rootTagName != "package")
+        throw Base::XMLBaseException("Malformed package.xml document: Root <package> group not found");
+
+    auto formatVersion = XMLString::parseInt(_dom->getAttribute(XUTF8Str("format").unicodeForm()));
+    switch (formatVersion) {
+    case 1:
+        parseVersion1(_dom);
+        break;
+    default:
+        throw Base::XMLBaseException("pacakge.xml format version is not supported by this version of FreeCAD");
+    }
+}
+
+Metadata::Metadata() : _dom(nullptr)
+{
+}
+
+Metadata::Metadata(const DOMNode* domNode, int format) : _dom(nullptr)
+{
+    auto element = dynamic_cast<const DOMElement*>(domNode);
+    if (element) {
+        switch (format) {
+        case 1:
+            parseVersion1(element);
+            break;
+        default:
+            throw Base::XMLBaseException("pacakge.xml format version is not supported by this version of FreeCAD");
+        }
+    }
+}
+
+Metadata::~Metadata()
+{
+}
+
+std::string Metadata::name() const
+{
+    return _name;
+}
+
+Meta::Version Metadata::version() const
+{
+    return _version;
+}
+
+std::string Metadata::description() const
+{
+    return _description;
+}
+
+std::vector<Meta::Contact> Metadata::maintainer() const
+{
+    return _maintainer;
+}
+
+std::vector<Meta::License> Metadata::license() const
+{
+    return _license;
+}
+
+std::vector<Meta::Url> Metadata::url() const
+{
+    return _url;
+}
+
+std::vector<Meta::Contact> Metadata::author() const
+{
+    return _author;
+}
+
+std::vector<Meta::Dependency> Metadata::depend() const
+{
+    return _depend;
+}
+
+std::vector<Meta::Dependency> Metadata::conflict() const
+{
+    return _conflict;
+}
+
+std::vector<Meta::Dependency> Metadata::replace() const
+{
+    return _replace;
+}
+
+std::vector<std::string> Metadata::tag() const
+{
+    return _tag;
+}
+
+fs::path Metadata::icon() const
+{
+    return _icon;
+}
+
+std::string Metadata::classname() const
+{
+    return _classname;
+}
+
+std::vector<fs::path> Metadata::file() const
+{
+    return _file;
+}
+
+Meta::Version App::Metadata::freecadmin() const
+{
+    return _freecadmin;
+}
+
+Meta::Version App::Metadata::freecadmax() const
+{
+    return _freecadmax;
+}
+
+std::multimap<std::string, Metadata> Metadata::content() const
+{
+    return _content;
+}
```

Comment:
Couldn't these functions return const values ?

Reply:
What would be the advantage?

Reply:
Hmmm... Maybe no one now we have C++11 as a minimum requirement...